### PR TITLE
Added iterator_adaptor, refactored pointer_iterator

### PR DESCRIPTION
**Pull request by JohanMabille**
This will make it easier to create other iterator adaptors.

### CODE REVIEW

File: include/sparrow/iterator.hpp [hunk 1]
```diff
@@ -480,9 +479,42 @@ namespace sparrow
...
+        
+        pointer_iterator() = default;
+        explicit pointer_iterator(const iterator_type& p)
+            : base_type(p)
+        {
+        }
+
+        template <class Iter>
+        requires std::same_as<Iter, std::remove_const_t<T>>
+        pointer_iterator(const Iter& iter)
```
Comment (jjerphan) [inline include/sparrow/iterator.hpp:515]: Is the template type parameter `Iter` used to ease the signature instead of using a `iterator_adaptor`?

**Review by jjerphan — APPROVED** [review]
LGTM modulo a suggestion and a question.

I do not see the how `pointer_iterator` is going to be used for now but I guess I will once adaptations are made.

**Review by Alex-PLACET — APPROVED** [review]

File: include/sparrow/iterator.hpp [hunk 2]
```diff
@@ -403,48 +403,47 @@ namespace sparrow
...
 
     /*
-     * @class pointer_iterator
-     * @brief iterator adaptor for pointers
+     * @class iterator_adaptor
+     * @brief generic iterator adaptor
      *
-     * pointer_iterator gives an iterator API to a pointer.
-     * @tparam T the pointer to adapt
+     * This class forward the calls to its "base" iterator, i.e.
```
Comment (jjerphan) [inline include/sparrow/iterator.hpp:409]:
```suggestion
     * This class forwards the calls to its "base" iterator, i.e.
```

File: include/sparrow/iterator.hpp [hunk 3]
```diff
@@ -480,9 +479,42 @@ namespace sparrow
...
+    /*
+     * @class pointer_iterator
+     * @brief iterator adaptor for pointers
+     *
+     * pointer_iterator gives an iterator API to a pointer.
+     * @tparam T the pointer to adapt
+     */
+    template <class T>
+    class pointer_iterator;
+
```
[the remark this answers is not in the document]
Comment (jjerphan) [reply include/sparrow/iterator.hpp:496]: I tested the removal locally and this is actually required for reasons I am trying to understand.

[the remark this answers is not in the document]
Comment (JohanMabille) [reply include/sparrow/iterator.hpp:496]: So actually this is the "pre-C++20" way to limit the definition of the pointer_iterator class to pointer template parameters:
- declare the class but do not implement it in the general case
- define the class for the T* specialization.

I will change to that to use concepts instead.

File: include/sparrow/iterator.hpp (hunk 1 above)
Comment (JohanMabille) [reply include/sparrow/iterator.hpp:515]: This allows to build a pointer_iterator<const T*> from a T*. But we're missing the `pointer_iterator<T*>` => `pointer_iterator<const T*>`

File: include/sparrow/iterator.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jjerphan) [reply include/sparrow/iterator.hpp:496]: As discussed IRL, we could define a concept to restrict the definition to pointers.

**Pull request merged**